ddtrace/tracer/stats.go

```diff
@@ -0,0 +1,362 @@
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016 Datadog, Inc.
+
+//go:generate msgp -unexported -marshal=false -o=stats_msgp.go -tests=false
+
+package tracer
+
+import (
+	"math/rand"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"google.golang.org/protobuf/proto"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/log"
+
+	"github.com/DataDog/datadog-go/statsd"
+	"github.com/DataDog/sketches-go/ddsketch"
+)
+
+// aggregableSpan holds necessary information about a span that can be used to
+// aggregate statistics in a bucket.
+type aggregableSpan struct {
+	// key specifies the aggregation key under which this span can be placed into
+	// grouped inside a bucket.
+	key aggregation
+
+	Start, Duration int64
+	Error           int32
+	TopLevel        bool
+}
+
+// defaultStatsBucketSize specifies the default span of time that will be
+// covered in one stats bucket.
+var defaultStatsBucketSize = (10 * time.Second).Nanoseconds()
+
+// concentrator aggregates and stores statistics on incoming spans in time buckets,
+// flushing them occasionally to the underlying transport located in the given
+// tracer config.
+type concentrator struct {
+	// In specifies the channel to be used for feeding data to the concentrator.
+	// In order for In to have a consumer, the concentrator must be started using
+	// a call to Start.
+	In chan *aggregableSpan
+
+	// mu guards below fields
+	mu sync.Mutex
+
+	// buckets maintains a set of buckets, where the map key represents
+	// the starting point in time of that bucket, in nanoseconds.
+	buckets map[int64]*rawBucket
+
+	// stopped reports whether the concentrator is stopped (when non-zero)
+	stopped uint64
+
+	wg         sync.WaitGroup // waits for any active goroutines
+	bufferLen  int            // maximum number of buckets buffered
+	bucketSize int64          // the size of a bucket in nanoseconds
+	oldestTs   int64          // any entry older than this will go into this bucket
+	stop       chan struct{}  // closing this channel triggers shutdown
+	cfg        *config        // tracer startup configuration
+}
+
+// defaultBufferLen represents the default buffer length; the number of bucket size
+// units used by the concentrator.
+const defaultBufferLen = 2
+
+// newConcentrator creates a new concentrator using the given tracer
+// configuration c. It creates buckets of bucketSize nanoseconds duration.
+func newConcentrator(c *config, bucketSize int64) *concentrator {
+	return &concentrator{
+		In:         make(chan *aggregableSpan, 10000),
+		bufferLen:  defaultBufferLen,
+		bucketSize: bucketSize,
+		stopped:    1,
+		oldestTs:   alignTs(time.Now().UnixNano(), bucketSize),
+		buckets:    make(map[int64]*rawBucket),
+		cfg:        c,
+	}
+}
+
+// alignTs returns the provided timestamp truncated to the bucket size.
+// It gives us the start time of the time bucket in which such timestamp falls.
+func alignTs(ts, bucketSize int64) int64 { return ts - ts%bucketSize }
```

Comment:
you can avoid aligning the timestamp and just add to the current live bucket (span just finished). And if it's older than 10s, flush and create a new bucket. There is no alignment constraint in the tracers.
span.End - currentBucket > 10s could be the trigger of a new bucket

Reply:
I guess this relates to the improvement in your other [comment](https://github.com/DataDog/dd-trace-go/pull/859/files#r617821126), right? Can we do that as a follow-up improvement?

Reply:
yes it can be done as a follow up, a good reason to remove this is to avoid synchronization: N tracers sending at the same time stats payload to the agent